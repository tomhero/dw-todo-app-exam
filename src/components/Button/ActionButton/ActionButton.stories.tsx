import React from 'react';
import { Meta, Story } from '@storybook/react';
import ActionButton, { ActionItem, ActionButtonProps } from './ActionButton';

export default {
  title: 'Components/Button/ActionButton',
  component: ActionButton,
  parameters: {
    docs: {
      description: {
        component: 'An Action Button Component',
      },
    },
  },
} as Meta;

const Template: Story<ActionButtonProps> = (args) => <ActionButton {...args} />;

const actions: ActionItem[] = [
  {
    label: 'Edit',
    action: 'edit',
  },
  {
    label: 'Delete',
    action: 'delete',
    extraClassName: 'dw-text-danger',
  },
];

export const Example = Template.bind({});
Example.args = {
  actions,
};
