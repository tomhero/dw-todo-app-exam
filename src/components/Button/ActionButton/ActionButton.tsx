import React, { useState } from 'react';

export type ActionItem = {
  label: string;
  action: string;
  extraClassName?: string;
};

export type ActionButtonProps = {
  id?: string;
  actions: ActionItem[];
  onClickAction: (action: string) => void;
};

const ActionButton = ({ id, actions, onClickAction }: ActionButtonProps) => {
  const [isOpenMenu, setIsOpenMenu] = useState(false);

  const selectHandler = (action: string) => {
    onClickAction(action);
    setIsOpenMenu(false);
  };

  return (
    <div id={id} className="dw-action-button">
      <img
        className="dw-action-button__dots"
        src="/icons/three-dots.svg"
        alt="action button dots"
      />
      <button
        onClick={() => setIsOpenMenu(!isOpenMenu)}
        className="dw-action-button__button"
        onBlur={() => {
          setTimeout(() => {
            setIsOpenMenu(false);
          }, 150);
        }}
      />
      <ul
        className={`dw-action-button__context-menu ${
          !isOpenMenu ? 'dw-action-button__context-menu--hide' : ''
        }`}
      >
        {actions.map(({ action, label, extraClassName }, index) => {
          const extraClass = extraClassName ? extraClassName : '';
          return (
            <li
              key={`${index}-${action}`}
              className={`dw-action-button__context-menu__item ${extraClass}`}
              onClick={() => selectHandler(action)}
            >
              {label}
            </li>
          );
        })}
      </ul>
    </div>
  );
};

export default ActionButton;
