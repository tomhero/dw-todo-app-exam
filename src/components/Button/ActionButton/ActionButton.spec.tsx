import { fireEvent, render } from '@test';
import ActionButton, { ActionButtonProps } from './ActionButton';

describe('Component testing with testing-library', () => {
  const defaultProps: ActionButtonProps = {
    actions: [
      { action: 'foo', label: 'Foo' },
      { action: 'bar', label: 'Bar' },
    ],
    onClickAction: jest.fn(),
  };

  beforeEach(() => {
    jest.useFakeTimers();
  });

  it('should render action button', () => {
    const mockingText = 'Foo';
    const hiddenClassName = 'dw-action-button__context-menu--hide';
    const { container, getByText, getByRole } = render(<ActionButton {...defaultProps} />);
    expect(container).toMatchSnapshot();
    expect(getByText(mockingText)).toBeDefined();
    expect(getByRole('list').classList.contains(hiddenClassName)).toBeTruthy();
  });

  it('can show context menu', () => {
    // Arrange
    const hiddenClassName = 'dw-action-button__context-menu--hide';

    // Act
    const { getByRole } = render(<ActionButton {...defaultProps} />);
    const btn = getByRole('button');
    fireEvent.click(btn);

    // Assert
    expect(getByRole('list').classList.contains(hiddenClassName)).toBeFalsy();
  });

  it('can select menu item', () => {
    // Arrange
    const hiddenClassName = 'dw-action-button__context-menu--hide';
    const mockSelectFn = jest.fn();
    const targetItem = 'bar';

    // Act
    const { getByRole } = render(<ActionButton {...defaultProps} onClickAction={mockSelectFn} />);
    const listItem = getByRole('list').getElementsByTagName('li')[1];
    fireEvent.click(listItem);

    // Assert
    expect(mockSelectFn).toHaveBeenCalledTimes(1);
    expect(mockSelectFn).toHaveBeenCalledWith(targetItem);
    expect(getByRole('list').classList.contains(hiddenClassName)).toBeTruthy();
  });

  it('should close context menu when blur event fired', () => {
    // Arrange
    const hiddenClassName = 'dw-action-button__context-menu--hide';

    // Act
    const { container, getByRole } = render(<ActionButton {...defaultProps} />);
    fireEvent.blur(container);

    // Assert
    expect(getByRole('list').classList.contains(hiddenClassName)).toBeTruthy();
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });
});
