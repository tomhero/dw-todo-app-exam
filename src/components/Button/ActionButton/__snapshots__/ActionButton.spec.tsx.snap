// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component testing with testing-library should render action button 1`] = `
<div>
  <div
    class="dw-action-button "
    tabindex="-1"
  >
    <img
      alt="action button dots"
      class="dw-action-button__dots"
      src="/icons/three-dots.svg"
    />
    <button
      class="dw-action-button__button"
    />
    <ul
      class="dw-action-button__context-menu dw-action-button__context-menu--hide"
    >
      <li
        class="dw-action-button__context-menu__item "
      >
        Foo
      </li>
      <li
        class="dw-action-button__context-menu__item "
      >
        Bar
      </li>
    </ul>
  </div>
</div>
`;
